### PR TITLE
Upgrade to github-api 1.114 and okhttp3

### PR DESCRIPTION
**Pull request by bitwiseman**
@KostyaSha 
This change updates github-plugi to use okhttp3. Okhttp 2.7.5 is seriously out of date and no longer supported. We should move to using okhttp3.

@KostyaSha @lanwen  

---
This change is [Reviewable](https://reviewable.io/reviews/jenkinsci/github-plugin/228)

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/github/internal/GitHubClientCacheOps.java [hunk 1]
```diff
@@ -94,6 +95,7 @@ public static Path getBaseCacheDir() {
      *
      * @param configs active server configs to exclude caches from cleanup
      */
+    @SuppressFBWarnings(value = "RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
```
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/internal/GitHubClientCacheOps.java:98]: hey, can you give more insights of why is this necessary?

File: src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java [hunk 2]
```diff
@@ -107,31 +106,15 @@ private Proxy getProxy(String apiUrl) {
      * @return connector to be used as backend for client
      */
     private OkHttpConnector connector(GitHubServerConfig config) {
-        OkHttpClient client = new OkHttpClient().setProxy(getProxy(defaultIfBlank(config.getApiUrl(), GITHUB_URL)));
+        OkHttpClient.Builder builder = BASECLIENT.newBuilder()
```
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java:109]: you can skip instance creation with builder creation directly via new OkHttpClient.Builder()

File: src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java [hunk 3]
```diff
@@ -111,7 +113,8 @@ public boolean apply(ITEM item) {
         if (item instanceof ParameterizedJobMixIn.ParameterizedJob) {
             ParameterizedJobMixIn.ParameterizedJob pJob = (ParameterizedJobMixIn.ParameterizedJob) item;
 
-            for (Trigger candidate : pJob.getTriggers().values()) {
+            Map<TriggerDescriptor, Trigger<?>> triggerMap = pJob.getTriggers();
```
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java:116]: can you give more details regarding this change?

File: pom.xml [hunk 4]
```diff
@@ -5,7 +5,7 @@
     <parent>
         <groupId>org.jenkins-ci.plugins</groupId>
         <artifactId>plugin</artifactId>
-        <version>3.54</version>
+        <version>4.2</version>
```
Comment (lanwen) [inline pom.xml:8]: 4 uses kotlin runtime, so I would rather stick with 3 to reduce this dependency which could be dangerous in a jenkins world

File: src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java (hunk 3 above)
Comment (KostyaSha) [reply src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java:116]: return type (generics) is changed on new core

File: pom.xml (hunk 4 above)
Comment (bitwiseman) [reply pom.xml:8]: @lanwen 
This is the plugin pom version number not the okhttp3 version number.

File: src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java (hunk 2 above)
Comment (bitwiseman) [reply src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java:109]: @lanwen 
The guidance is to have a single client and then create new clients from that client using `newBuilder()`.

File: src/main/java/org/jenkinsci/plugins/github/internal/GitHubClientCacheOps.java (hunk 1 above)
Comment (bitwiseman) [reply src/main/java/org/jenkinsci/plugins/github/internal/GitHubClientCacheOps.java:98]: @lanwen 
Spotbugs continues to have issues with properly identifying the the null check automatically generated when using [try with resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) as something that cannot be fixed by the person writing the code.  It reports a problem with [line 109](https://github.com/jenkinsci/github-plugin/pull/228/files#diff-92e6819e3ec96cf3b77dec0a1aece0deR109) which isn't even changed in this PR.

File: pom.xml (hunk 4 above)
Comment (lanwen) [reply pom.xml:8]: lol, sorry, wrong look :)

File: pom.xml [hunk 5]
```diff
@@ -47,10 +47,11 @@
...
     <properties>
-        <jenkins.version>2.60.3</jenkins.version>
+        <jenkins.version>2.164.3</jenkins.version>
         <release.skipTests>false</release.skipTests>
         <maven.javadoc.skip>true</maven.javadoc.skip>
         <findbugs-maven-plugin.version>3.0.4</findbugs-maven-plugin.version>
+        <hamcrest.version>2.2</hamcrest.version>
         <concurrency>1</concurrency>
         <java.level>8</java.level>
         <workflow.version>1.14.2</workflow.version>
```
Comment (oleg-nenashev) [inline pom.xml:57]: Would be great to switch to https://github.com/jenkinsci/bom in a later pull request. This version of Pipeline is ancient, and I here we are not testing with versions being really used with the target Jenkins core

File: pom.xml [hunk 6]
```diff
@@ -162,26 +150,31 @@
...
+            <version>${hamcrest.version}</version>
             <scope>test</scope>
         </dependency>
-
+      
         <dependency>
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
-            <version>4.12</version>
+            <version>4.13</version>
```
Comment (oleg-nenashev) [inline pom.xml:177]: Parent POM provides it

```suggestion
```

File: pom.xml [hunk 7]
```diff
@@ -306,6 +299,16 @@
 
     </dependencies>
 
+    <dependencyManagement>
```
Comment (oleg-nenashev) [inline pom.xml:302]: Why is it needed? BOM supplies it
https://github.com/jenkinsci/jenkins/blob/master/bom/pom.xml#L170-L174

Comment (bitwiseman) [reply pom.xml:302]: Not using bom.

File: pom.xml (hunk 5 above)
Comment (bitwiseman) [reply pom.xml:57]: Agree bom is a good move, but that is a separate change.

**Comment by KostyaSha** [comment]
probably, ok. Is it done?

**Comment by bitwiseman** [comment]
@KostyaSha 
Yes this is done.

**Comment by KostyaSha** [comment]
let's try
